src/librustc/ty/inhabitedness.rs

```diff
@@ -0,0 +1,264 @@
+// Copyright 2012-2015 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use std::mem;
+use rustc_data_structures::small_vec::SmallVec;
+use syntax::ast::CRATE_NODE_ID;
+use util::nodemap::FxHashSet;
+use ty::context::TyCtxt;
+use ty::{AdtDef, VariantDef, FieldDef, TyS};
+use ty::{DefId, Substs};
+use ty::{AdtKind, Visibility, DefIdTree};
+use ty::TypeVariants::*;
+
+/// Represents a set of DefIds closed under the ancestor relation. That is, if
```

Comment:
What "ancestor" relation? These comments are quite clear, but they don't say much about how this is used. From reading the code, this is my hypothesis:

- These def-ids represent a set of modules. 
- When we have some ADT type like `Foo<T, U, !>`, where some fields/variants have type !, we will compute a set of modules that can see that field and hence can deduce that this ADT is uninhabited, and store those in a forest. 

Seem right? If so, I think it'd be great to talk a bit about how this type is used here.

I'd probably also make this inhabitedness/mod.rs and put the forest code into forest.rs. Small modules ftw! :)